Sources/NIOOpenSSL/SSLPKCS12Bundle.swift

```diff
@@ -0,0 +1,218 @@
+    ///
+    /// If the bundle does have a passphrase, call `init(file:passphrase:)` instead.
+    ///
+    /// - parameters:
+    ///     - file: The path to the PKCS#12 bundle on disk.
+    public init(file: String) throws {
+        try self.init(file: file, passphrase: nil as [UInt8]?)
+    }
+
+    /// Create a `OpenSSLPKCS12Bundle` from the given bytes in memory,
+    /// assuming it has no passphrase.
+    ///
+    /// If the bundle does have a passphrase, call `init(buffer:passphrase:)` instead.
+    ///
+    /// - parameters:
+    ///     - buffer: The bytes of the PKCS#12 bundle.
+    public init(buffer: [UInt8]) throws {
+        try self.init(buffer: buffer, passphrase: nil as [UInt8]?)
+    }
+}
+
+
+internal extension Collection where Element == UInt8 {
+    /// Provides a contiguous copy of the bytes of this collection in a heap-allocated
+    /// memory region that is locked into memory (that is, which can never be backed by a file),
+    /// and which will be scrubbed and freed after use.
+    ///
+    /// This method should be used when it is necessary to take a secure copy of a collection of
+    /// bytes. Its implementation relies on OpenSSL directly.
+    internal func withSecureBytes<T>(_ block: (UnsafePointer<Int8>) throws -> T) throws -> T {
+        // We need to allocate some memory and prevent it being swapped to disk while we use it.
+        // For that reason we use mlock.
+        // Note that here we use UnsafePointer and UnsafeBufferPointer. Ideally we'd just use
+        // the buffer pointer, but 4.0 doesn't have the APIs we want: specifically, .allocate
+        // and .withMemoryRebound(to:) are only added to the buffer pointers in 4.1.
+        let bufferSize = Int(self.count) + 1
+        let ptr = UnsafeMutablePointer<UInt8>.allocate(capacity: bufferSize)
+        let bufferPtr = UnsafeMutableBufferPointer(start: ptr, count: bufferSize)
+        defer {
+            ptr.deallocate()
+        }
+
+        if mlock(bufferPtr.baseAddress!, bufferPtr.count) != 0 {
+            throw IOError(errnoCode: errno, function: "mlock")
+        }
+        defer {
+            // If munlock fails take out the process.
+            let rc = munlock(bufferPtr.baseAddress!, bufferPtr.count)
+            precondition(rc == 0, "munlock failed with \(rc), errno \(errno)")
+        }
+
+        let (_, nextIndex) = bufferPtr.initialize(from: self)
+        defer {
+            // We use OPENSSL_cleanse here because the compiler can't optimize this away.
+            // .initialize(repeating: 0) can be, and empirically is, optimized away, bzero
+            // is deprecated, memset_s is not well supported cross-platform, and memset-to-zero
+            // is famously easily optimised away. This is our best bet.
+            OPENSSL_cleanse(bufferPtr.baseAddress!, bufferPtr.count)
+        }
+        assert(nextIndex == (bufferPtr.endIndex - 1))
+
+        // Add a null terminator.
+        bufferPtr[nextIndex] = 0
```

Comment:
I think this is technically wrong as you're writing to unintialised memory here. I think it should be `bufferPtr.advanced(by: nextIndex).initialize(to: 0)`.

I realise that `UInt8` is a primitive type that doesn't have any ref counts but imagine you'd have saved a type that needs ref counts. In that case this would non-deterministically crash. Because `bufferPtr[nextIndex] = newValue` would first read out the old value, decrement the ref count, the increment the ref count of `newValue` and then write it in there. `initialize` on the other hand doesn't touch the ref count of the memory but just increases `newValue`'s ref count and then writes it into the memory (which then becomes initialised).

Yes, `UInt8` doesn't have a ref count of course but still the same rules apply. Memory needs to be 1) allocated 2) initialised 3) written as often as the user pleases 4) deinitialised 5) deallocated

You're doing (1), most of (2) (except for the last byte), (3) and (5). So parts of (1) and (4) are missing